src/main/scala/io/iohk/ethereum/blockchain/sync/FastSyncStateHandler.scala

```diff
@@ -0,0 +1,446 @@
+    import syncConfig.{fastSyncBlockValidationN => N}
+    if (header.number <= syncState.safeDownloadTarget) {
+      discardLastBlocks(header.number, N)
+      if (header.number >= syncState.targetBlock.number) {
+        InitTargetBlockUpdate(LastBlockValidationFailed(BlackListCommand(peer, "failed validation of final headers")))
+      } else {
+        ContinueSyncing(Some(BlackListCommand(peer, reason)))
+      }
+    } else {
+      ContinueSyncing(Some(BlackListCommand(peer, reason)))
+    }
+  }
+}
+
+// scalastyle:off
+object FastSyncStateHandler {
+  sealed trait HashType {
+    def v: ByteString
+  }
+
+  case class StateMptNodeHash(v: ByteString) extends HashType
+  case class ContractStorageMptNodeHash(v: ByteString) extends HashType
+  case class EvmCodeHash(v: ByteString) extends HashType
+  case class StorageRootHash(v: ByteString) extends HashType
+
+  sealed abstract class HeaderProcessingResult{
+    def newState: SyncState
+  }
+  case class  HeadersProcessingFinished(newState: SyncState)                         extends HeaderProcessingResult
+  case class  ParentDifficultyNotFound(header:BlockHeader, newState: SyncState)      extends HeaderProcessingResult
+  case class  ValidationFailed(header:BlockHeader, peer: Peer, newState: SyncState)  extends HeaderProcessingResult
+  case class  ImportedTargetBlock(newState: SyncState)                               extends HeaderProcessingResult
+
+  case class  ValidationSucces(header:BlockHeader, newState: SyncState)
+
+
+  sealed abstract class FinalBlockProcessingResult
+  case object ImportedLastBlock         extends FinalBlockProcessingResult
+  case class LastBlockValidationFailed(blackListCommand: BlackListCommand) extends FinalBlockProcessingResult
+
+
+  case class BlackListCommand(peer: Peer, reason: String)
+  sealed abstract class FastSyncCommand
+  case class ContinueSyncing(peerToBlacklist: Option[BlackListCommand])   extends FastSyncCommand
+  case class InitTargetBlockUpdate(reason: FinalBlockProcessingResult)    extends FastSyncCommand
+  case object AbortFastSync                                               extends FastSyncCommand
+  case class ContTargetBlockUpdate(reason: FinalBlockProcessingResult)    extends FastSyncCommand
+  case object BackToSyncing                                               extends FastSyncCommand
+  case class AskForNewTarget(reason: FinalBlockProcessingResult)          extends FastSyncCommand
+  case class RequestReceipts(from: Peer, receiptsToGet: Seq[ByteString])  extends FastSyncCommand
+  case class RequestBodies(from: Peer, bodiesToGet: Seq[ByteString])      extends FastSyncCommand
+  case class RequestNodes(from: Peer,
+                          nodesToGet: Seq[HashType],
+                          mptNodesToGet: Seq[HashType],
+                          nonMptNodesToGet: Seq[HashType])                 extends FastSyncCommand
+  case class RequestHeaders(from: Peer, starting: BigInt, number: BigInt)  extends FastSyncCommand
+
+
```

Comment:
consider here one enter instead of two, same in 59-60, 112-113, 151-152, 422-423, 427-428